lib/ansible/vars/manager.py

```diff
@@ -196,14 +196,10 @@ def _combine_and_track(data, new_data, source):
             basedirs = [self._loader.get_basedir()]
 
         if play:
-            if not C.DEFAULT_PRIVATE_ROLE_VARS:
-                # first we compile any vars specified in defaults/main.yml
-                # for all roles within the specified play
-                for role in play.get_roles():
-                    # role from roles or include_role+public or import_role and completed
-                    if not role.from_include or role.public or (role.static and role._completed.get(to_text(host), False)):
-                        all_vars = _combine_and_track(all_vars, role.get_default_vars(), "role '%s' defaults" % role.name)
-
+            for role in play.get_roles():
+                # role from roles if explicitly public or if import_role and config defaults to public
+                if role.public or (role.public is None and role.static and not C.DEFAULT_PRIVATE_ROLE_VARS):
```

Comment:
This is a change in behavior for include_role too, now  `C.DEFAULT_PRIVATE_ROLE_VARS` is disregarded. No idea if this would blow up in CI, but this just backs out the completion condition for `import_role` exports (and adjusts a few tests to be compatible with pre-https://github.com/ansible/ansible/pull/69040 behavior):
```diff
diff --git a/lib/ansible/vars/manager.py b/lib/ansible/vars/manager.py
index b6adad34b7..7190163f3e 100644
--- a/lib/ansible/vars/manager.py
+++ b/lib/ansible/vars/manager.py
@@ -200,8 +200,8 @@ class VariableManager:
                 # first we compile any vars specified in defaults/main.yml
                 # for all roles within the specified play
                 for role in play.get_roles():
-                    # role from roles or include_role+public or import_role and completed
-                    if not role.from_include or role.public or (role.static and role._completed.get(to_text(host), False)):
+                    # role from roles or include_role+public or import_role
+                    if not role.from_include or role.public or role.static:
                         all_vars = _combine_and_track(all_vars, role.get_default_vars(), "role '%s' defaults" % role.name)
 
         if task:
@@ -390,7 +390,7 @@ class VariableManager:
             # unless the user has disabled this via a config option
             if not C.DEFAULT_PRIVATE_ROLE_VARS:
                 for role in play.get_roles():
-                    if not role.from_include or role.public or (role.static and role._completed.get(to_text(host), False)):
+                    if not role.from_include or role.public or role.static:
                         all_vars = _combine_and_track(all_vars, role.get_vars(include_params=False, only_exports=True), "role '%s' exported vars" % role.name)
 
         # next, we merge in the vars from the role, which will specifically
diff --git a/test/integration/targets/include_import/public_exposure/no_bleeding.yml b/test/integration/targets/include_import/public_exposure/no_bleeding.yml
index bce5d90e1a..b9db713290 100644
--- a/test/integration/targets/include_import/public_exposure/no_bleeding.yml
+++ b/test/integration/targets/include_import/public_exposure/no_bleeding.yml
@@ -5,8 +5,8 @@
     - name: Static imports should expose vars at parse time, not at execution time
       assert:
         that:
-          - static_defaults_var is not defined
-          - static_vars_var is not defined
+          - static_defaults_var == 'static_defaults'
+          - static_vars_var == 'static_vars'
     - import_role:
         name: static
     - assert:
diff --git a/test/integration/targets/include_import/public_exposure/playbook.yml b/test/integration/targets/include_import/public_exposure/playbook.yml
index ea9641974c..11735e7731 100644
--- a/test/integration/targets/include_import/public_exposure/playbook.yml
+++ b/test/integration/targets/include_import/public_exposure/playbook.yml
@@ -10,16 +10,12 @@
     - name: Static imports should expose vars at parse time, not at execution time
       assert:
         that:
-          - static_defaults_var is not defined
-          - static_vars_var is not defined
-          - static_task_var is not defined
+          - static_defaults_var == 'static_defaults'
+          - static_vars_var == 'static_vars'
     - import_role:
         name: static
     - assert:
         that:
-          - static_vars_var is defined
-          - static_tasks_var is defined
-          - static_defaults_var is defined
           - static_tasks_var == 'static_tasks'
           - static_defaults_var == 'static_defaults'
           - static_vars_var == 'static_vars'
diff --git a/test/integration/targets/roles/vars_scope.yml b/test/integration/targets/roles/vars_scope.yml
index 256723a230..802cb0fabc 100644
--- a/test/integration/targets/roles/vars_scope.yml
+++ b/test/integration/targets/roles/vars_scope.yml
@@ -64,6 +64,24 @@
           roles_and_role_vars: role_vars
           role_vars_only: role_vars
 
+- name: play baseline (no roles)
+  hosts: localhost
+  gather_facts: false
+  vars_files:
+    - vars/play.yml
+  tasks:
+    - include_tasks: roles/vars_scope/tasks/check_vars.yml
+      vars:
+        defined:
+          play_and_import: play
+          play_and_import_and_role_vars: play
+          play_and_include: play
+          play_and_include_and_role_vars: play
+          play_and_roles: play
+          play_and_roles_and_role_vars: play
+          play_and_role_vars: play
+          play_only: play
+
 - name: play and import
   hosts: localhost
   gather_facts: false
@@ -81,6 +99,16 @@
           play_and_roles_and_role_vars: play
           play_and_role_vars: play
           play_only: play
+          default_only: default
+          import_and_role_vars: role_vars
+          include_and_role_vars: role_vars
+          play_and_import_and_role_vars: role_vars
+          play_and_role_vars: role_vars
+          play_and_role_vars_and_role_vars: role_vars
+          play_and_include_and_role_vars: role_vars
+          play_and_roles_and_role_vars: role_vars
+          roles_and_role_vars: role_vars
+          role_vars_only: role_vars
 
     - name: static import
       import_role:
```
no change needed to this section, and we could still allow `import_role` to configure `public` in 2.17+.

Reply:
> I still think explicit public on role import/include should > config option, I know this is a change, but it seems a sensible one.

This trips me up every time I look at this bit of code. :smile: After I look at the config description I'm always appeased, but +1 for changing it (could mention in the docs/changelog).